RecoTracker/MkFit/plugins/createCMS2017.cc

```diff
@@ -0,0 +1,372 @@
+      const LayerInfo &tecn1 = trk_info.m_layers[54];
+
+      const LayerInfo &tec_first = z_dir_pos ? tecp1 : tecn1;
+
+      // If a track hits outer barrel ... it is in the barrel (for central, "outgoing" tracks).
+      // This is also true for cyl-cow.
+      // Better check is: hits outer TIB, misses inner TEC (but is +-z dependant).
+      // XXXX Calculate z ... then check is inside or less that first EC z.
+      // There are a lot of tracks that go through that crack.
+
+      // XXXX trying a fix for low pT tracks that are in barrel after half circle
+      float maxR = S.maxReachRadius();
+      float z_at_maxr;
+
+      bool can_reach_outer_brl = S.canReachRadius(outer_brl.m_rout);
+      float z_at_outer_brl;
+      bool misses_first_tec;
+      if (can_reach_outer_brl) {
+        z_at_outer_brl = S.zAtR(outer_brl.m_rout);
+        if (z_dir_pos)
+          misses_first_tec = z_at_outer_brl < tec_first.m_zmin;
+        else
+          misses_first_tec = z_at_outer_brl > tec_first.m_zmax;
+      } else {
+        z_at_maxr = S.zAtR(maxR);
+        if (z_dir_pos)
+          misses_first_tec = z_at_maxr < tec_first.m_zmin;
+        else
+          misses_first_tec = z_at_maxr > tec_first.m_zmax;
+      }
+
+      if (/*can_reach_outer_brl &&*/ misses_first_tec)
+      // outer_brl.is_within_z_limits(S.zAtR(outer_brl.r_mean())))
+      {
+        reg = TrackerInfo::Reg_Barrel;
+      } else {
+        // This should be a list of layers
+        // CMS, first tib, tob: 4, 10
+
+        if ((S.canReachRadius(tib1.m_rin) && tib1.is_within_z_limits(S.zAtR(tib1.m_rin))) ||
+            (S.canReachRadius(tob1.m_rin) && tob1.is_within_z_limits(S.zAtR(tob1.m_rin)))) {
+          // transition region ... we are still hitting barrel layers
+
+          reg = z_dir_pos ? TrackerInfo::Reg_Transition_Pos : TrackerInfo::Reg_Transition_Neg;
+        } else {
+          // endcap ... no barrel layers will be hit anymore.
+
+          reg = z_dir_pos ? TrackerInfo::Reg_Endcap_Pos : TrackerInfo::Reg_Endcap_Neg;
+        }
+      }
+
+      part.m_region[i] = reg;
+      part.m_sort_score[i] = 5.0f * (reg - 2) + eta;
+    }
+  }
+}  // namespace
+
+namespace mkfit {
+  void createCMS2017(TrackerInfo &ti, IterationsInfo &ii, bool verbose) {
+    // TODO: these writes to global variables need to be removed
+    Config::nTotalLayers = 18 + 2 * 27;
+
+    Config::useCMSGeom = true;
+
+    Config::finding_requires_propagation_to_hit_pos = true;
+    Config::finding_inter_layer_pflags = PropagationFlags(PF_use_param_b_field | PF_apply_material);
+    Config::finding_intra_layer_pflags = PropagationFlags(PF_none);
+    Config::backward_fit_pflags = PropagationFlags(PF_use_param_b_field | PF_apply_material);
+    Config::forward_fit_pflags = PropagationFlags(PF_use_param_b_field | PF_apply_material);
+    Config::seed_fit_pflags = PropagationFlags(PF_none);
+    Config::pca_prop_pflags = PropagationFlags(PF_none);
```

Comment:
These writes to global variables need to (eventually) be removed (the sooner the better).

Reply:
Agreed. nTotalLayers is used by validation, I think. useCMSGeom should definitely be removed everywhere (and optionally replaced by some more "functional" config/params) ... and finding/propagation flags should go to IterationConfig.